libraries/hotstuff/chain_pacemaker.cpp

```diff
@@ -112,9 +112,31 @@ namespace eosio { namespace hotstuff {
       return _chain->is_builtin_activated( builtin_protocol_feature_t::instant_finality );
    }
 
-   void chain_pacemaker::get_state( finalizer_state& fs ) const {
-      if (enabled())
-         _qc_chain.get_state( fs ); // get_state() takes scare of finer-grained synchronization internally
+   void chain_pacemaker::get_state(finalizer_state& fs, bool force) const {
+      if (! enabled())
```

Comment:
This is still accessing chain state from non-main thread. I assume you are going to address that in part two of this PR.

Reply:
I was thinking about solving _chain-> in a separate PR, then close #1541 when both PRs are in.